File: scripts/generate-dataset.mjs
// Generate src/data/fullDataset.ts from the canonical pronunciation_dataset_v1_0.json
// Run: node scripts/generate-dataset.mjs <path-to-json>
import { readFileSync, writeFileSync } from 'fs';
import { resolve } from 'path';

const jsonPath = process.argv[2];
if (!jsonPath) {
  console.error("Usage: node scripts/generate-dataset.mjs <path-to-dataset.json>");
  process.exit(1);
}

console.log(`Reading ${jsonPath}...`);
const full = JSON.parse(readFileSync(jsonPath, 'utf-8'));
let allEntries = full.entries || full;

if (!Array.isArray(allEntries)) {
  console.error("Expected JSON to have an 'entries' array or be a top-level array");
  process.exit(1);
}

// Sort by lessonOrder, then id for stable output
allEntries.sort((a, b) => (a.lessonOrder - b.lessonOrder) || (a.id - b.id));

console.log(`Loaded ${allEntries.length} entries`);

// Sanity check: unique IDs
const ids = allEntries.map(e => e.id);
const uniqueIds = new Set(ids);
if (uniqueIds.size !== ids.length) {
  console.error(`Duplicate IDs found: ${ids.length - uniqueIds.size} dupes`);
  process.exit(1);
}

// Sanity check: required fields
const required = ['id', 'displayName', 'sentence', 'audioPromptText', 'level',
                  'lessonOrder', 'module', 'practiceStage', 'skillType',
                  'difficultyScore', 'instruction', 'tags', 'targetSounds', 'speechaceIntegration'];
for (const e of allEntries) {
  for (const field of required) {
    if (!(field in e)) {
      console.error(`Entry ${e.id}: missing required field '${field}'`);
      process.exit(1);
    }
  }
}

const header = `// AUTO-GENERATED from pronunciation_dataset_v1_0.json
// Do not edit by hand — re-run: npm run generate-dataset
// Source: ${jsonPath}
// Generated: ${new Date().toISOString()}
// Entry count: ${allEntries.length}

import { DatasetEntry } from '../types';

export const FULL_DATASET: DatasetEntry[] = `;

const ts = header + JSON.stringify(allEntries, null, 2) + ';\n';
const outPath = resolve('src/data/fullDataset.ts');
writeFileSync(outPath, ts, 'utf-8');

console.log(`Wrote ${allEntries.length} entries to ${outPath}`);
console.log(`File size: ${(ts.length / 1024).toFixed(1)} KB`);

File: scripts/generate-dataset.py
#!/usr/bin/env python3
"""Generate src/data/fullDataset.ts from the canonical dataset JSON.
Usage: python scripts/generate-dataset.py <path-to-dataset.json>
"""
import json, os, sys
from datetime import datetime, timezone

if len(sys.argv) < 2:
    print("Usage: python scripts/generate-dataset.py <path-to-dataset.json>")
    sys.exit(1)

json_path = sys.argv[1]
print(f"Reading {json_path}...")

with open(json_path, 'r', encoding='utf-8') as f:
    full = json.load(f)

all_entries = full.get('entries', full) if isinstance(full, dict) else full

if not isinstance(all_entries, list):
    print("Expected JSON to have an 'entries' array or be a top-level array")
    sys.exit(1)

# Sort by lessonOrder, then id
all_entries.sort(key=lambda e: (e.get('lessonOrder', 0), e.get('id', 0)))

print(f"Loaded {len(all_entries)} entries")

# Sanity: unique IDs
ids = [e['id'] for e in all_entries]
if len(set(ids)) != len(ids):
    dupes = [i for i in set(ids) if ids.count(i) > 1]
    print(f"Duplicate IDs found: {dupes}")
    sys.exit(1)

# Sanity: required fields
required = ['id', 'displayName', 'sentence', 'audioPromptText', 'level',
            'lessonOrder', 'module', 'practiceStage', 'skillType',
            'difficultyScore', 'instruction', 'tags', 'targetSounds', 'speechaceIntegration']
for e in all_entries:
    for field in required:
        if field not in e:
            print(f"Entry {e.get('id')}: missing required field '{field}'")
            sys.exit(1)

# Build TypeScript
header = f"""// AUTO-GENERATED from pronunciation_dataset_v1_0.json
// Do not edit by hand — re-run: npm run generate-dataset
// Source: {json_path}
// Generated: {datetime.now(timezone.utc).isoformat()}
// Entry count: {len(all_entries)}

import {{ DatasetEntry }} from '../types';

export const FULL_DATASET: DatasetEntry[] = """

ts = header + json.dumps(all_entries, indent=2, ensure_ascii=False) + ';\n'

out_path = os.path.join('src', 'data', 'fullDataset.ts')
os.makedirs(os.path.dirname(out_path), exist_ok=True)
with open(out_path, 'w', encoding='utf-8') as f:
    f.write(ts)

print(f"Wrote {len(all_entries)} entries to {out_path}")
print(f"File size: {len(ts) / 1024:.1f} KB")
